### DISCUSSION TITLE
page.add_highlight_annot(start=pointa, stop=pointb) not working

### DISCUSSION THREAD

**Discussion by kalimer00** [pymupdf/PyMuPDF#2168] [Looking for help] [answered]
Hello,

i'm trying to highlight a text in a pdf from a start word "pointa" to an end word "pointb" but it wont work it will mark all the text on the page Maybe some one could help me (pleas) and figure out what i'm doing wrong.

output:
![output](https://user-images.githubusercontent.com/32965279/211621525-b58d9861-1ff0-41b7-bafa-f1f62aecb9b5.png)

`import fitz
import config
class FindTextblock:

        doc = fitz.open(config.fpath_int+"/test_source.pdf")
        for page in doc:
            wordlist = page.get_text("words")
            wordlist.sort(key=lambda w: (w[1], w[0]))  # sort vertical, then horizontal
            pointa = None
            pointb = None
            for w in wordlist:
                if w[4] == "Therapievorschlag":  # choose a word to start marking
                    pointa = fitz.Point(w[:2])  # top left of work rectangle
                    break
            for w in wordlist:
                if w[4] == "EU-Verordnung":  # some word for stopping the marking
                    pointb = fitz.Point(w[2:4])  # bottom right of word rectangle
                    break

            page.add_highlight_annot(start=pointa, stop=pointb)

        doc.save("test.pdf")

FindTextblock()`

test file:
[test_source.pdf](https://github.com/pymupdf/PyMuPDF/files/10385320/test_source.pdf)

Thank you

**Comment by JorjMcKie** [comment]
@kalimer00 - welcome and hello again!
Thanks for using this channel. I will have a look.

**Reply by JorjMcKie** [reply]
Made minor adjustments to your script, now it does work perfectly:
![grafik](https://user-images.githubusercontent.com/8290722/211624646-0132846c-17f6-4ca7-a3b3-4627150c8de1.png)
The modified script:
```python
import fitz

doc = fitz.open("test_source.pdf")
for page in doc:
    wordlist = page.get_text("words", sort=True)  # has built-in parameter
    pointa = None
    pointb = None
    for w in wordlist:
        if w[4] == "Therapievorschlag":  # choose a word to start marking
            pointa = fitz.Point(w[:2])  # top left of work rectangle
            break
    for w in wordlist:
        if w[4] == "EU-Verordnung":  # some word for stopping the marking
            pointb = fitz.Point(w[2:4])  # bottom right of word rectangle
            break

    if pointa != None and pointb != None:  # guard against hick ups
        page.add_highlight_annot(start=pointa, stop=pointb)

doc.save("test.pdf")
```

**Comment by JorjMcKie** [comment, marked as answer]
This script is even a bit simpler:
```python
import fitz

doc = fitz.open("test_source.pdf")
for page in doc:
    rl1 = page.search_for("Therapievorschlag")
    if rl1 == []:  # not on page
        continue
    rl2 = page.search_for("EU-Verordnung")
    if rl2 == []:  # not on page
        continue
    pointa = rl1[0].tl
    pointb = rl2[0].br
    page.add_highlight_annot(start=pointa, stop=pointb)

doc.save("test.pdf")
```

**Reply by kalimer00** [reply]
wow, that was fast. thank you so much, it works like a charm.

**Reply by JorjMcKie** [reply]
always welcome!

**Reply by kalimer00** [reply]
hello, the code works fine when im execute it on a single file, but if i loop through a folder with pdf files it will damage the output files and when i try to open them they will throw the error massage “A number is out of range”

![Screenshot 2023-01-13 114746](https://user-images.githubusercontent.com/32965279/212304086-409a7e51-f09a-4788-8da5-ccfa7a681503.png)

`import fitz, os, config

def mark_text_block(path, start_word, end_word):

    number = 0
    # doc = fitz.open(config.fpath_int+"/test_source.pdf")
    res = os.listdir(path)
    for file in res:
        if file.endswith(".pdf"):
            print("File in folder: "+file)
            print("Path to file: " +path+file)
            print("-------------------------")
            doc = fitz.open(path+file, filetype="pdf")
            for page in doc:
                wordlist = page.get_text("words", sort=True)  # has built-in parameter
                pointa = None
                pointb = None
                for w in wordlist:
                    if w[4] == start_word:  # choose a word to start marking
                        pointa = fitz.Point(w[:2])  # top left of work rectangle
                        break
                for w in wordlist:
                    if w[4] == end_word:  # some word for stopping the marking
                        pointb = fitz.Point(w[2:4])  # bottom right of word rectangle
                        break
                if pointa != None and pointb != None:  # guard against hick ups
                    page.add_highlight_annot(start=pointa, stop=pointb)

            number = number + 1
            doc.save(str(number)+"_"+file)

mark_text_block(config.fpath_int, config.start_text_2, config.end_text)`

**Comment by JorjMcKie** [comment]
There is no way to help without reproducible, sorry. Maybe to clarify a few questions:
* Can you reproduce with a directory with only one file?
* Does the same file work in standalone, but no longer when in a directory?
* How do to protect against multiple changes of the same file? If you prefix the filename with a number, this may happen in a second go, right? Try to write in a separate directory, or choose a naming pattern for new files such that they can be safely skipped. E.g. `doc.save(doc.name.replace(".pdf", "-marked.pdf"))` and skip files ending like this on input.

**Reply by JorjMcKie** [reply]
Also make sure, that `pointb.y > pointa.y`.

**Reply by kalimer00** [reply]
Hello,

thanks for your help.
I have checked the questions you mentioned and figured out that the PDF file is the problem. 
The file is generated with a Java program. If I edit the right text field in Acrobat Reader on the page where the annotation goes and then save the file, I can run the script without error and the annotation is applied.

I can't upload the files because of the sensitive patient data but I could email you the files if you want to have a look.

**Reply by JorjMcKie** [reply]
Oh please do so!
There always is a chance to improve ...

**Reply by JorjMcKie** [reply]
Received your example PDFs and rechecked, thank you. Everything works fine - could not detect any problem.

**Reply by kalimer00** [reply]
You could apply the annotation on both PDF files without the error massage "a number is out of range" after opening?

**Reply by JorjMcKie** [reply]
No! Instead I found a problem: because of how this PDF is created, corners of infinite text rectangles are stored in the annotation definition. Those are the number Adobe Acrobat is complaining about.
Other PDF viewer are more tolerant and simply ignore obvious nonsense.
Will be fixed in the next version.

**Reply by kalimer00** [reply]
Aaaah ok perfekt, thanks for your help
